utils/analysis/filter_test.go

```diff
@@ -196,14 +196,67 @@ func TestFilterAnalysisRunsToDelete(t *testing.T) {
 		deletedRS,
 		nil,
 	}
-	filteredArs := FilterAnalysisRunsToDelete(ars, olderRSs)
+	filteredArs := FilterAnalysisRunsToDelete(ars, olderRSs, 4, 4)
 	assert.Len(t, filteredArs, 3)
 	assert.NotContains(t, filteredArs, arNoDeletion)
 	assert.Contains(t, filteredArs, arWithNoPodHash)
 	assert.Contains(t, filteredArs, arWithDeletedRS)
 	assert.Contains(t, filteredArs, arWithNoMatchingRS)
 }
 
+func TestFilterAnalysisRunsToDeleteByLimit(t *testing.T) {
+	rs := func(podHash string) *appsv1.ReplicaSet {
+		return &appsv1.ReplicaSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{v1alpha1.DefaultRolloutUniqueLabelKey: podHash},
+			},
+		}
+	}
+	ar := func(podHash string, phase v1alpha1.AnalysisPhase) *v1alpha1.AnalysisRun {
+		return &v1alpha1.AnalysisRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels:            map[string]string{v1alpha1.DefaultRolloutUniqueLabelKey: podHash},
+				CreationTimestamp: metav1.Now(),
+			},
+			Status: v1alpha1.AnalysisRunStatus{
+				Phase: phase,
+			},
+		}
+	}
+
+	arS0 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arS1 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arS2 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arS3 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arS4 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arF0 := ar("a", v1alpha1.AnalysisPhaseFailed)
+	arF1 := ar("a", v1alpha1.AnalysisPhaseFailed)
+	arF2 := ar("a", v1alpha1.AnalysisPhaseFailed)
```

Comment:
Can we mix the Successful and Failed analysis to make the test more robust? Thanks.

Reply:
Do you mean shuffle the order of ARs?

Reply:
Yep

Reply:
done